File: web/src/station/migration/skuFacts.ts
// Client-side port of api/skufacts.py — stable SKU fact IDs + deterministic
// fact-reference extraction. Used for the local-sample path and for media-node
// asset dependencies; the backend remains the source of truth when reachable.

const UNIT = '(?:°\\s*[cf]|℃|℉|ml|cl|l|oz|lbs?|cm|mm|kg|mah|hz|inch(?:es)?|ft|%)';
const NUM_TOKEN_RE = new RegExp(`-?\\d+(?:\\.\\d+)?\\s*${UNIT}?`, 'gi');
const ASCII_WORD_RE = /[a-z]{4,}/g;
const CJK_RUN_RE = /[一-鿿]{2,}/g;
const FACT_ID_RE = /^fact-(\d+)$/;

const STOPWORDS = new Set([
  'with', 'that', 'this', 'from', 'your', 'have', 'will', 'into', 'than',
  'them', 'then', 'they', 'and', 'the', 'for', 'not',
]);

export function parseSkuFacts(productName: string, points: string): Record<string, string> {
  const facts: Record<string, string> = {};
  const name = (productName || '').trim();
  if (name) facts.name = name;
  let idx = 0;
  for (const raw of (points || '').split('\n')) {
    const line = raw.trim();
    if (!line) continue;
    idx += 1;
    facts[`fact-${idx}`] = line;
  }
  return facts;
}

export function factSortKey(id: string): [number, number, string] {
  if (id === 'name') return [0, 0, ''];
  const m = FACT_ID_RE.exec(id);
  if (m) return [1, Number(m[1]), ''];
  return [2, 0, id];
}

export function salientTokens(text: string): Set<string> {
  const tokens = new Set<string>();
  if (!text) return tokens;
  const lowered = text.toLowerCase();
  const compact = lowered.replace(/\s+/g, '');

  for (const m of lowered.matchAll(NUM_TOKEN_RE)) {
    const tok = m[0].replace(/\s+/g, '');
    if (!/^-?\d/.test(tok)) continue;
    tokens.add(tok);
    const num = /^-?\d+(?:\.\d+)?/.exec(tok);
    if (num) {
      const digits = num[0].replace('-', '').split('.')[0];
      if (digits.length >= 2) tokens.add(num[0].replace('-', ''));
    }
  }
  for (const m of lowered.matchAll(ASCII_WORD_RE)) {
    if (!STOPWORDS.has(m[0])) tokens.add(m[0]);
  }
  for (const m of text.matchAll(CJK_RUN_RE)) {
    if (m[0].length >= 2) tokens.add(m[0]);
  }
  if (compact.includes('bpa')) tokens.add('bpa');

  return new Set([...tokens].filter(t => t.length >= 2));
}

function isCjk(token: string): boolean {
  return !!token && token.charCodeAt(0) >= 0x4e00 && token.charCodeAt(0) <= 0x9fff;
}

export function computeFactRefs(text: string, facts: Record<string, string>): string[] {
  if (!text) return [];
  const compact = text.toLowerCase().replace(/\s+/g, '');
  const refs = new Set<string>();
  for (const [factId, value] of Object.entries(facts)) {
    for (const token of salientTokens(value)) {
      const hit = isCjk(token) ? text.includes(token) : compact.includes(token);
      if (hit) {
        refs.add(factId);
        break;
      }
    }
  }
  return [...refs].sort((a, b) => {
    const ka = factSortKey(a);
    const kb = factSortKey(b);
    return ka[0] - kb[0] || ka[1] - kb[1] || ka[2].localeCompare(kb[2]);
  });
}

export function diffFacts(
  before: Record<string, string>,
  after: Record<string, string>,
): { added: string[]; removed: string[]; changed: string[] } {
  const added = Object.keys(after).filter(k => !(k in before));
  const removed = Object.keys(before).filter(k => !(k in after));
  const changed = Object.keys(after).filter(
    k => k in before && before[k].trim() !== after[k].trim(),
  );
  const sort = (xs: string[]) =>
    xs.sort((a, b) => {
      const ka = factSortKey(a);
      const kb = factSortKey(b);
      return ka[0] - kb[0] || ka[1] - kb[1] || ka[2].localeCompare(kb[2]);
    });
  return { added: sort(added), removed: sort(removed), changed: sort(changed) };
}

File: api/skufacts.py
"""Stable SKU fact IDs and deterministic fact-reference extraction.

A SKU's facts are its product name (``name``) plus one entry per non-empty
selling-point line (``fact-1``, ``fact-2`` …). Generated listing fields record
which fact IDs they depend on (``factRefs``) so a later fact change can be
turned into a precise blast radius instead of "regenerate everything".

Everything here is deterministic string work — no model, no network.
"""

from __future__ import annotations

import hashlib
import re
from typing import Iterable

_STOPWORDS = {
    "with", "that", "this", "from", "your", "have", "will", "into", "than",
    "them", "then", "they", "and", "the", "for", "not", "但是", "而且", "以及",
    "适合", "可用", "可以",
}

# number + optional known unit. The unit is a whitelist so a number that runs
# straight into the next word ("300mlfoldable") can't swallow letters.
_UNIT = r"(?:°\s*[cf]|℃|℉|ml|cl|l|oz|lbs?|cm|mm|kg|mah|hz|inch(?:es)?|ft|%)"
_NUM_TOKEN_RE = re.compile(r"-?\d+(?:\.\d+)?\s*" + _UNIT + r"?", re.IGNORECASE)
_ASCII_WORD_RE = re.compile(r"[a-z]{4,}")
_CJK_RUN_RE = re.compile(r"[一-鿿]{2,}")
_FACT_ID_RE = re.compile(r"^fact-(\d+)$")


def parse_sku_facts(product_name: str, points: str) -> dict[str, str]:
    """``{"name": ..., "fact-1": <line 1>, "fact-2": <line 2>, ...}``.

    Selling-point lines are split on newlines; blank lines are skipped but do
    *not* shift the numbering of the lines that follow — numbering is by
    surviving line order, which is stable for a given input.
    """
    facts: dict[str, str] = {}
    name = (product_name or "").strip()
    if name:
        facts["name"] = name
    idx = 0
    for raw in (points or "").split("\n"):
        line = raw.strip()
        if not line:
            continue
        idx += 1
        facts[f"fact-{idx}"] = line
    return facts


def fact_sort_key(fact_id: str) -> tuple[int, int, str]:
    if fact_id == "name":
        return (0, 0, "")
    m = _FACT_ID_RE.match(fact_id)
    if m:
        return (1, int(m.group(1)), "")
    return (2, 0, fact_id)


def salient_tokens(text: str) -> set[str]:
    """Distinctive tokens of a fact: measurements, long numbers, meaningful
    words, and short CJK n-grams. Used only for substring matching."""
    if not text:
        return set()
    lowered = text.lower()
    compact = lowered.replace(" ", "")
    tokens: set[str] = set()

    for m in _NUM_TOKEN_RE.finditer(lowered):
        tok = re.sub(r"\s+", "", m.group(0))
        if not re.match(r"-?\d", tok):
            continue
        tokens.add(tok)
        num = re.match(r"-?\d+(?:\.\d+)?", tok)
        if num:
            digits = num.group(0).lstrip("-").split(".")[0]
            if len(digits) >= 2:
                tokens.add(num.group(0).lstrip("-"))

    for m in _ASCII_WORD_RE.finditer(lowered):
        word = m.group(0)
        if word not in _STOPWORDS:
            tokens.add(word)

    # Full CJK runs only (no sub-grams): matching must be a real shared phrase,
    # not an incidental two-character overlap ("防漏盖" vs "防漏测试").
    for m in _CJK_RUN_RE.finditer(text):
        run = m.group(0)
        if run not in _STOPWORDS and len(run) >= 2:
            tokens.add(run)

    if "bpa" in compact:
        tokens.add("bpa")

    return {t for t in tokens if len(t) >= 2}


def _is_cjk(token: str) -> bool:
    return bool(token) and "一" <= token[0] <= "鿿"


def compute_fact_refs(text: str, facts: dict[str, str]) -> list[str]:
    """Fact IDs whose distinctive tokens appear in *text*. Sorted, deduped."""
    if not text:
        return []
    lowered_compact = text.lower().replace(" ", "")
    refs: set[str] = set()
    for fact_id, value in facts.items():
        for token in salient_tokens(value):
            hit = token in text if _is_cjk(token) else token in lowered_compact
            if hit:
                refs.add(fact_id)
                break
    return sorted(refs, key=fact_sort_key)


def validate_fact_refs(refs: Iterable[str], facts: dict[str, str]) -> list[str]:
    """Drop unknown / malformed IDs; keep order stable."""
    valid = set(facts)
    seen: set[str] = set()
    out: list[str] = []
    for ref in refs or []:
        ref = str(ref).strip()
        if ref in valid and ref not in seen:
            seen.add(ref)
            out.append(ref)
    return sorted(out, key=fact_sort_key)


def hash_fact(value: str) -> str:
    return hashlib.sha1(value.strip().encode("utf-8")).hexdigest()[:12]


def fact_hashes(facts: dict[str, str]) -> dict[str, str]:
    return {fid: hash_fact(val) for fid, val in facts.items()}


def sku_revision_hash(facts: dict[str, str]) -> str:
    joined = "\n".join(f"{fid}={facts[fid]}" for fid in sorted(facts, key=fact_sort_key))
    return hashlib.sha1(joined.encode("utf-8")).hexdigest()[:16]


def diff_facts(old: dict[str, str], new: dict[str, str]) -> dict[str, list[str]]:
    """IDs added / removed / changed between two fact maps."""
    old_h = fact_hashes(old)
    new_h = fact_hashes(new)
    added = [k for k in new if k not in old]
    removed = [k for k in old if k not in new]
    changed = [k for k in new if k in old and old_h[k] != new_h[k]]
    return {
        "added": sorted(added, key=fact_sort_key),
        "removed": sorted(removed, key=fact_sort_key),
        "changed": sorted(changed, key=fact_sort_key),
    }
